deduplication/services.py

```diff
@@ -187,19 +191,87 @@ def _is_model_column(model: Union[Type[ExtendableModel], Type[HistoryModel]], co
+            if current_value != value:
+                is_instance_updated = True
+                if value:
+                    json_ext[key] = value
+                else:
+                    del json_ext[key]
+        instance.json_ext = json_ext
+        if is_instance_updated:
+            instance.save(username=user.username)
+
+
+@transaction.atomic
+def merge_duplicate_beneficiaries(task_data, user_id):
+    individual_fields = {"first_name", "last_name", "dob"}
+    beneficiary_fields = {"status"}
+
+    user = User.objects.get(id=user_id)
+    # additional_resolve_data is a key in task__json_ext that stores data selected by user during resolving a task
+    additional_resolve_data = task_data.get("json_ext", {}).get("additional_resolve_data", {})
+    merge_data = list(additional_resolve_data.values())[0]
+    field_values = merge_data.get('values', {})
+    beneficiary_ids = merge_data.get("beneficiaryIds", [])
+
+    oldest_record_beneficiary = Beneficiary.objects.filter(id__in=beneficiary_ids).order_by('date_created').first()
+    if oldest_record_beneficiary is None:
+        return  # No beneficiaries found to merge
+
+    beneficiaries_to_delete = Beneficiary.objects.filter(id__in=beneficiary_ids).exclude(
+        id=oldest_record_beneficiary.id)
+
+    individual_kwargs = {}
+    beneficiary_kwargs = {}
+    json_kwargs = {}
+
+    for key, value in field_values.items():
+        if key in individual_fields:
+            individual_kwargs[key] = value
+        elif key in beneficiary_fields:
+            beneficiary_kwargs[key] = value
+        else:
+            json_kwargs[key] = value
+
+    _update_instance_if_different_value(oldest_record_beneficiary.individual, individual_kwargs, user)
+    _update_instance_if_different_value(oldest_record_beneficiary, beneficiary_kwargs, user)
+    _update_instance_json_ext_if_different_value(oldest_record_beneficiary, json_kwargs, user)
+
+    # needs to be done this way because overridden .delete() does not work on qs
+    for beneficiary_to_delete in beneficiaries_to_delete:
+        beneficiary_to_delete.delete(username=user.username)
+
+
+def on_deduplication_task_complete_service_handler():
     def func(**kwargs):
         try:
             result = kwargs.get('result', {})
             task = result['data']['task']
-            business_event = task['business_event']
-            if result and result['success'] \
-                    and task['status'] == Task.Status.COMPLETED:
-                operation = business_event.split(".")[1]
-                if hasattr(service, operation):
-                    user = User.objects.get(id=result['data']['user']['id'])
-                    data = task['data']['incoming_data']
-                    getattr(service(user), operation)(data)
+            user_id = result['data']['user']["id"]
+            if result and result['success'] and task['status'] == Task.Status.COMPLETED:
+                merge_duplicate_beneficiaries(task, user_id)
```

Comment:
This 'complete_task' handler is missing task source check which means it will try to run `merge_duplicate_beneficiaries` for every successfully completed task.

Reply:
DONE